# NumPy Structured Array Operations
## Introduction
NumPy is a Python library used for numerical computing and array-based data processing. 
In this assignment, the provided `students.csv` dataset is used to perform basic NumPy operations.

The dataset contains information about students through three columns:
- `name` – student's name
- `age` – student's age
- `height` – student's height

The main objective is to use basic NumPy features to create and manipulate a structured array.
The following operations will be performed:
1. Load and inspect the provided dataset.
2. Create a structured NumPy array with the fields `name`, `age`, and `height`.
3. Access and print all values of the `name` field.
4. Update the age of the first individual to 30.
5. Add a new record for Nadine Smith.
6. Display the results and findings from each operation.

The operations are performed using basic NumPy features so that the relationship between the dataset and NumPy array operations can be clearly understood.

## 1. Loading the Dataset

The provided dataset is stored in a CSV file named `students.csv`.

NumPy provides functions for reading data from files. Here, `np.genfromtxt()` is used to load the CSV file.

The `names=True` argument allows NumPy to read the first row of the CSV file as column names.

The `dtype=None` argument allows NumPy to automatically determine suitable data types for the columns.

After loading the dataset, the data and its column names are displayed.

In [15]:
import numpy as np
# Load the provided CSV dataset using NumPy
data = np.genfromtxt(
    'students.csv',
    delimiter=',',
    names=True,
    dtype=None,
    encoding='utf-8'
)
print("Dataset loaded successfully.")
print("\nColumn names:")
print(data.dtype.names)

print("\nFirst five records:")
print(data[:5])

Dataset loaded successfully.

Column names:
('name', 'age', 'height')

First five records:
[('Anisha Regmi', 21, 6.3) ('Amit Ghimire', 18, 6.1)
 ('Sujal Tamang', 19, 5.4) ('Amit Rai', 18, 5.5)
 ('Sabina Acharya', 25, 6.1)]


### Finding

The `students.csv` dataset was successfully loaded using NumPy.
The dataset contains three fields:
- `name`
- `age`
- `height`

The first five records show that the dataset already contains the exact types of information required for the structured array assignment.

## 2. Basic NumPy Dataset Inspection

Before performing the required operations, some basic NumPy properties are used to understand the dataset.

The number of records, field names, and individual fields are inspected.

This helps us understand the structure of the data before manipulating it.

In [16]:
# Number of records
print("Number of records:", len(data))

# Field names
print("Fields:", data.dtype.names)

# Access the age field
print("\nAge values:")
print(data['age'][:10])

# Access the height field
print("\nHeight values:")
print(data['height'][:10])

Number of records: 120
Fields: ('name', 'age', 'height')

Age values:
[21 18 19 18 25 23 21 22 22 25]

Height values:
[6.3 6.1 5.4 5.5 6.1 5.4 5.2 5.7 6.2 5.9]


### Finding

The dataset contains 120 student records.

The dataset has three fields: `name`, `age`, and `height`.

Using NumPy field selection, individual columns can be accessed directly. 
For example, `data['age']` returns the age values and `data['height']` returns the height values.

This demonstrates basic NumPy structured-array field access.

## 3. Part A — Creating a Structured Array

The assignment requires a structured array containing five individuals with the fields:

- `name` – string
- `age` – integer
- `height` – float

Since the provided dataset already contains these three fields, the first five records from the loaded dataset are selected.

A structured NumPy data type is defined manually so that each field has an appropriate data type.

The following data types are used:

- `U30` – Unicode string with space for up to 30 characters
- `i4` – integer
- `f8` – floating-point number

In [17]:
# Define the required structured array data types
data_type = [
    ('name', 'U30'),
    ('age', 'i4'),
    ('height', 'f8')
]

# Select the first five records from the loaded dataset
sample_data = data[:5]

# Create the structured NumPy array
students = np.array(sample_data, dtype=data_type)

print("Structured Array:")
print(students)

Structured Array:
[('Anisha Regmi', 21, 6.3) ('Amit Ghimire', 18, 6.1)
 ('Sujal Tamang', 19, 5.4) ('Amit Rai', 18, 5.5)
 ('Sabina Acharya', 25, 6.1)]


### Finding

The first five records from the provided dataset were successfully converted into a structured NumPy array.

The structured array contains the three required fields:

`name`, `age`, and `height`.

Each field has an appropriate data type, allowing different types of information to be stored in the same NumPy array.

In [18]:
print("Structured Array Data Type:")
print(students.dtype)

Structured Array Data Type:
[('name', '<U30'), ('age', '<i4'), ('height', '<f8')]


### Finding

The output confirms that the structured array contains three different data types:

- `name` is stored as a string.
- `age` is stored as an integer.
- `height` is stored as a floating-point value.

This confirms that the structured array has been created correctly.

## 4. Part B — Accessing the `name` Field

A specific field in a NumPy structured array can be accessed using its field name.

The expression `students['name']` selects only the `name` field.

This allows all student names to be displayed without accessing the age or height fields.

In [19]:
# Access all values from the name field
names = students['name']

print("All names:")
print(names)

All names:
['Anisha Regmi' 'Amit Ghimire' 'Sujal Tamang' 'Amit Rai' 'Sabina Acharya']


### Finding

The `name` field was successfully accessed using NumPy field indexing.

The five names from the selected dataset records were displayed without accessing the other fields.

## 5. Part C — Updating the Age of the First Individual

NumPy arrays use zero-based indexing, so the first record is located at index `0`.

The first student's original age is accessed using:

`students[0]['age']`

The assignment requires the age of the first individual to be changed to `30`.

In [20]:
# Display the original age
print("Original age:", students[0]['age'])

# Update the age
students[0]['age'] = 30

# Display the updated age
print("Updated age:", students[0]['age'])

# Display the complete updated record
print("\nUpdated first record:")
print(students[0])

Original age: 21
Updated age: 30

Updated first record:
('Anisha Regmi', 30, 6.3)


### Finding

The first student's age was successfully updated.

The original age was 21 and it was changed to 30.

The student's name and height remained unchanged, demonstrating that a specific field of a structured NumPy array can be modified independently.

## 6. Part D — Adding a New Record

The assignment requires a new record with the following values:

- Name: ` Nadine Smith `
- Age: `25`
- Height: `5.9`

NumPy arrays have a fixed size, so the new record is first created using the same structured data type.

`np.append()` is then used to create a new array containing the existing records and the new record.

In [21]:
# Create the new record using the same structured data type
new_student = np.array(
    [(' Nadine Smith ', 25, 5.9)],
    dtype=data_type
)

# Add the new record
students = np.append(students, new_student)

print("Final Structured Array:")
print(students)

Final Structured Array:
[('Anisha Regmi', 30, 6.3) ('Amit Ghimire', 18, 6.1)
 ('Sujal Tamang', 19, 5.4) ('Amit Rai', 18, 5.5)
 ('Sabina Acharya', 25, 6.1) (' Nadine Smith ', 25, 5.9)]


### Finding

The new record for Nadine Smith was successfully added to the structured array.

The array now contains six records: the original five selected from the dataset and the newly added record.

## 7. Final Result

The final structured array is displayed below after completing all required operations.

The first individual's age has been changed to 30, and Nadine Smith has been added as a new record.

In [22]:
print("Final Student Records:\n")

for student in students:
    print(
        "Name:", student['name'],
        "| Age:", student['age'],
        "| Height:", student['height']
    )

Final Student Records:

Name: Anisha Regmi | Age: 30 | Height: 6.3
Name: Amit Ghimire | Age: 18 | Height: 6.1
Name: Sujal Tamang | Age: 19 | Height: 5.4
Name: Amit Rai | Age: 18 | Height: 5.5
Name: Sabina Acharya | Age: 25 | Height: 6.1
Name:  Nadine Smith  | Age: 25 | Height: 5.9


## 8. Overall Findings

The NumPy operations produced the following findings:

1. The provided `students.csv` dataset was successfully loaded using NumPy.
2. The dataset contains 120 records and three fields: `name`, `age`, and `height`.
3. The first five records were successfully converted into a structured NumPy array.
4. The `name` field was accessed independently using NumPy field indexing.
5. The age of the first individual was changed from 21 to 30.
6. A new record for Nadine Smith was successfully added.
7. The final structured array contains six records.

These operations demonstrate basic NumPy features such as reading CSV data, structured arrays, field access, indexing, updating values, and appending records.

In [23]:
print("===== FINAL FINDINGS =====")

print("Original dataset records:", len(data))
print("Records in final array:", len(students))

print("\nFirst individual's final age:")
print(students[0]['age'])

print("\nNames in final array:")
print(students['name'])

print("\nNewly added record:")
print(students[-1])

===== FINAL FINDINGS =====
Original dataset records: 120
Records in final array: 6

First individual's final age:
30

Names in final array:
['Anisha Regmi' 'Amit Ghimire' 'Sujal Tamang' 'Amit Rai' 'Sabina Acharya'
 ' Nadine Smith ']

Newly added record:
(' Nadine Smith ', 25, 5.9)
